File: Dart/Creational/builder.dart
/// Enum to define the pizza sizes
enum PizzaSize { small, medium, large }

/// Enum to define all pizza toppings
enum Toppings {
  mozarella,
  groundBeef,
  sausage,
  pepperoni,
  onion,
  mushrooms,
  greenPapers,
  tomatoeSauce,
  tangyTomatoePaste,
  cheddar,
  fetaCheese,
  blackOlives,
  redOnion,
}

/// Interface defines how to build a Pizza
abstract interface class PizzaBuilder {
  PizzaBuilder({
    required PizzaSize pizzaSize,
    required List<Toppings> pizzaToppings,
  })  : _pizzaSize = pizzaSize,
        _pizzaToppings = pizzaToppings;

  final PizzaSize _pizzaSize;
  final List<Toppings> _pizzaToppings;

  /// Returns the pizza after build
  String getPizza() => '''
Kind: ${this.runtimeType}
Size: $_pizzaSize
Toppings: $_pizzaToppings
''';
}

/// Concrete Chicago pizza buider
class ChicagoPizza extends PizzaBuilder {
  ChicagoPizza({required PizzaSize pizzaSize})
      : super(pizzaSize: pizzaSize, pizzaToppings: [
          Toppings.mozarella,
          Toppings.groundBeef,
          Toppings.sausage,
          Toppings.pepperoni,
          Toppings.onion,
          Toppings.mushrooms,
          Toppings.greenPapers,
        ]);
}

/// Concrete New York pizza buider
class NewYorkPizza extends PizzaBuilder {
  NewYorkPizza({required PizzaSize pizzaSize})
      : super(pizzaSize: pizzaSize, pizzaToppings: [
          Toppings.mozarella,
          Toppings.tomatoeSauce,
        ]);
}

/// Concrete Greek pizza buider
class GreekPizza extends PizzaBuilder {
  GreekPizza({required PizzaSize pizzaSize})
      : super(pizzaSize: pizzaSize, pizzaToppings: [
          Toppings.tangyTomatoePaste,
          Toppings.mozarella,
          Toppings.cheddar,
          Toppings.fetaCheese,
          Toppings.blackOlives,
          Toppings.redOnion,
        ]);
}

/// Director to build all kinds of pizzas
class Director {
  String buildPizza(PizzaBuilder builder) => builder.getPizza();
}

void main() {
  final director = Director();

  final pizza1 = director.buildPizza(ChicagoPizza(pizzaSize: PizzaSize.small));
  final pizza2 = director.buildPizza(NewYorkPizza(pizzaSize: PizzaSize.medium));
  final pizza3 = director.buildPizza(GreekPizza(pizzaSize: PizzaSize.large));

  print(pizza1);
  print(pizza2);
  print(pizza3);

  /// Here is the result:
/*
Kind: ChicagoPizza
Size: PizzaSize.small
Toppings: [Toppings.mozarella, Toppings.groundBeef, Toppings.sausage, Toppings.pepperoni, Toppings.onion, Toppings.mushrooms, Toppings.greenPapers]

Kind: NewYorkPizza
Size: PizzaSize.medium
Toppings: [Toppings.mozarella, Toppings.tomatoeSauce]

Kind: GreekPizza
Size: PizzaSize.large
Toppings: [Toppings.tangyTomatoePaste, Toppings.mozarella, Toppings.cheddar, Toppings.fetaCheese, Toppings.blackOlives, Toppings.redOnion]
*/
}

File: Swift/Creational/builder.swift
//
//  builder.swift
//
//

// Enum to define the pizza sizes
enum PizzaSize { case small, medium, large }

// Enum to define all pizza toppings
enum Toppings { case mozzarella, groundBeef, pepperoni, olives, mushrooms, tomatoes, fetaCheese }

// Protocol defines how to build a pizza
protocol PizzaBuilder {
    var pizzaSize: PizzaSize { get }
    var pizzaToppings: [Toppings] { get }
}

// Default getPizza() shared by all builders
extension PizzaBuilder {
    func getPizza() -> String {
        let kind = String(describing: type(of: self))
        let size = String(describing: pizzaSize)
        let toppings = pizzaToppings.map { String(describing: $0) }.joined(separator: ", ")
        
        return "Kind: \(kind)\nSize: \(size)\nToppings: \(toppings)"
    }
}

// Concrete Chicago pizza builder
struct ChicagoPizza: PizzaBuilder {
    let pizzaSize: PizzaSize
    let pizzaToppings: [Toppings]
    
    init(pizzaSize: PizzaSize = .medium,
         pizzaToppings: [Toppings] = [.mozzarella, .groundBeef, .mushrooms]) {
        self.pizzaSize = pizzaSize
        self.pizzaToppings = pizzaToppings
    }
}

// Concrete New York pizza builder
struct NewYorkPizza: PizzaBuilder {
    let pizzaSize: PizzaSize
    let pizzaToppings: [Toppings]
    
    init(pizzaSize: PizzaSize = .medium,
         pizzaToppings: [Toppings] = [.mozzarella, .pepperoni, .olives]) {
        self.pizzaSize = pizzaSize
        self.pizzaToppings = pizzaToppings
    }
}

// Concrete Greek pizza builder
struct GreekPizza: PizzaBuilder {
    let pizzaSize: PizzaSize
    let pizzaToppings: [Toppings]
    
    init(pizzaSize: PizzaSize = .medium,
         pizzaToppings: [Toppings] = [.fetaCheese, .olives, .tomatoes]) {
        self.pizzaSize = pizzaSize
        self.pizzaToppings = pizzaToppings
    }
}

// Director to build all kinds of pizzas
class Director {
    func buildPizza(_ builder: any PizzaBuilder) -> String { builder.getPizza() }
}

// Demo
let director = Director()

let chicago = director.buildPizza(ChicagoPizza(pizzaSize: .large))
let newYork = director.buildPizza(NewYorkPizza())
let greek = director.buildPizza(GreekPizza(pizzaSize: .small))

print(chicago)
print("\n")
print(newYork)
print("\n")
print(greek)

// Expected output:
// Kind: ChicagoPizza
// Size: large
// Toppings: mozzarella, groundBeef, mushrooms
//
// Kind: NewYorkPizza
// Size: medium
// Toppings: mozzarella, pepperoni, olives
//
// Kind: GreekPizza
// Size: small
// Toppings: fetaCheese, olives, tomatoes
